pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/invertedindex/InvertedIndexHandler.java

```diff
@@ -45,17 +45,41 @@
 
   private final File _indexDir;
   private final SegmentMetadata _segmentMetadata;
+  private final SegmentDirectory.Reader _segmentReader;
   private final SegmentDirectory.Writer _segmentWriter;
   private final HashSet<String> _columnsToAddIdx;
 
   public InvertedIndexHandler(File indexDir, SegmentMetadata segmentMetadata, IndexLoadingConfig indexLoadingConfig,
       SegmentDirectory.Writer segmentWriter) {
+    this(indexDir, segmentMetadata, indexLoadingConfig, null, segmentWriter);
+  }
+
+  public InvertedIndexHandler(SegmentMetadata segmentMetadata, IndexLoadingConfig indexLoadingConfig,
+      SegmentDirectory.Reader segmentReader) {
+    this(null, segmentMetadata, indexLoadingConfig, segmentReader, null);
+  }
+
+  private InvertedIndexHandler(File indexDir, SegmentMetadata segmentMetadata, IndexLoadingConfig indexLoadingConfig,
```

Comment:
When do we need to instantiate the handler with both segment reader and writer ? I see IndexHandlerFactory separately using the reader (for checking if preprocess is needed) and writer(for creating new)

Reply:
I don't think there is a need for that. I extracted this private constructor to make the two public ones shorter. I thought about to pass in SegmentDirectory, but the pattern of using SegmentDirectory.Writer is to accumulate as many writes as possible before closing for efficiency. It's costly to open/close Writer repetitively.

Reply:
refactored, and now segment reader/writer is passed to updateIndicies() and needUpdateIndices(), instead of to constructor.